types/index.d.ts

```diff
@@ -0,0 +1,494 @@
+
+export function setContext<C extends object>(
+  props: C,
+): SagaGenerator<void, SetContextEffect<C>>;
+
+export function getContext<T = unknown>(
+  prop: string,
+): SagaGenerator<T, GetContextEffect>;
+
+export function delay<T = true>(
+  ms: number,
+  val?: T,
+): SagaGenerator<T, CallEffect<T>>;
+
+export function throttle<P extends ActionPattern>(
+  ms: number,
+  pattern: P,
+  worker: (action: ActionMatchingPattern<P>) => any,
+): SagaGenerator<never, ForkEffect<never>>;
+export function throttle<
+  P extends ActionPattern,
+  Fn extends (...args: any[]) => any
+>(
+  ms: number,
+  pattern: P,
+  worker: Fn,
+  ...args: HelperWorkerParameters<ActionMatchingPattern<P>, Fn>
+): SagaGenerator<never, ForkEffect<never>>;
+export function throttle<A extends Action>(
+  ms: number,
+  pattern: ActionPattern<A>,
+  worker: (action: A) => any,
+): SagaGenerator<never, ForkEffect<never>>;
+export function throttle<A extends Action, Fn extends (...args: any[]) => any>(
+  ms: number,
+  pattern: ActionPattern<A>,
+  worker: Fn,
+  ...args: HelperWorkerParameters<A, Fn>
+): SagaGenerator<never, ForkEffect<never>>;
+export function throttle<T>(
+  ms: number,
+  channel: TakeableChannel<T>,
+  worker: (item: T) => any,
+): SagaGenerator<never, ForkEffect<never>>;
+export function throttle<T, Fn extends (...args: any[]) => any>(
+  ms: number,
+  channel: TakeableChannel<T>,
+  worker: Fn,
+  ...args: HelperWorkerParameters<T, Fn>
+): SagaGenerator<never, ForkEffect<never>>;
+
+export function debounce<P extends ActionPattern>(
+  ms: number,
+  pattern: P,
+  worker: (action: ActionMatchingPattern<P>) => any,
+): SagaGenerator<never, ForkEffect<never>>;
+export function debounce<
+  P extends ActionPattern,
+  Fn extends (...args: any[]) => any
+>(
+  ms: number,
+  pattern: P,
+  worker: Fn,
+  ...args: HelperWorkerParameters<ActionMatchingPattern<P>, Fn>
+): SagaGenerator<never, ForkEffect<never>>;
+export function debounce<A extends Action>(
+  ms: number,
+  pattern: ActionPattern<A>,
+  worker: (action: A) => any,
+): SagaGenerator<never, ForkEffect<never>>;
+export function debounce<A extends Action, Fn extends (...args: any[]) => any>(
+  ms: number,
+  pattern: ActionPattern<A>,
+  worker: Fn,
+  ...args: HelperWorkerParameters<A, Fn>
+): SagaGenerator<never, ForkEffect<never>>;
+
+export function debounce<T>(
+  ms: number,
+  channel: TakeableChannel<T>,
+  worker: (item: T) => any,
+): SagaGenerator<never, ForkEffect<never>>;
+export function debounce<T, Fn extends (...args: any[]) => any>(
+  ms: number,
+  channel: TakeableChannel<T>,
+  worker: Fn,
+  ...args: HelperWorkerParameters<T, Fn>
+): SagaGenerator<never, ForkEffect<never>>;
+
+export function retry<Fn extends (...args: any[]) => any>(
+  maxTries: number,
+  delayLength: number,
+  fn: Fn,
+  ...args: Parameters<Fn>
+): SagaGenerator<SagaReturnType<Fn>, CallEffect<SagaReturnType<Fn>>>;
+
+type EffectReturnType<T> = T extends CallEffect
```

Comment:
Any idea how to improve this heuristic?

Reply:
I've updated it in https://github.com/agiledigital/typed-redux-saga/commit/2aebeb1ccc464280986e6a4ed253ad5955ed4f10#diff-88bec6beae84369c0376b56b3bb88fe1R468 to include `SagaGenerator`